### PR TITLE
New Mod to allow raising your stats naturally through the progression…

### PR DESCRIPTION
**Pull request by DeNarr**
… of skills.

### CODE REVIEW

**Comment by Marrim** [comment]
Should have diminishing returns, I think.
It'd be much harder to raise your strength from 15 to 16, than it would be from 5 to 6.

**Comment by DeNarr** [comment]
Well, that stats are based off skill levels. So it would indeed be harder to raise from 15 to 16, since it is harder to raise skills at higher levels.

**Comment by Marrim** [comment]
At higher stat levels, not higher skill levels.

**Comment by DeNarr** [comment]
Ah, that's the thing. You can't start with higher stats with this mod. Says so right in the description. It would indeed be broken if you could start with higher stats, and then have them increase from there. If you want to start with higher stats, you'll need to instead start with higher skill points for the relevant stat.

**Comment by kevingranade** [comment]
Each instance of accumulating skills onto the bonus can be something like:
```stat_bonus += min(3, skill_level / 3)```
I'm not sure if lua has a min, if not:
```
If skill_level / 3 > 3
then
    stat_bonus += skill_level / 3
else
    stat_bonus += 3
```
Also I suspect the mutation boni will either be applied twice or overwrite
these values.

**Comment by DeNarr** [comment]
Wouldn't that mean there would be a minimum of a +3 bonus for each skill? With the possibility of it being even greater if the skills were 12 or above? I was limiting it to a max of +3 per skill, since getting above 9 on a skill tends to not matter anyways.

As for the mutations, they shouldn't get applies more than once. The bonus is recalculated each time it is called.

**Comment by kevingranade** [comment]
No, min will return the smaller of its arguments, so if the stat / 3 is
smaller that will be the result, but if stat / 3 is greater than 3, it will
return 3 instead.
Your callback is invoked once a day, but the mutation effects are called
once a turn, and your code doesn't do anything to keep them from getting
applied.  I haven't looked at the code for it, but that's my suspicion.

**Comment by DeNarr** [comment]
Oh, I was talking about your if/then statement.

Hmm, I saw that the setting of the stats was over writing the mutation bonus, which is why I took the time to put them all in individually.

**Comment by kevingranade** [comment]
oop, yea I got that backwards, but I think you can see what I mean.

**Comment by Coolthulhu** [comment]
Ignores mutations for some reason. Does `has_trait` check mutations too or only inherent traits?

**Comment by DeNarr** [comment]
@Coolthulhu That was my fault. I hadn't realized that while the traits go from BENDY to BENDY2, they go from STR_UP to STR_UP_2. I missed the second underscore, and when I tested, I was only using the first level of the mutation. It should be fixed now.

File: data/mods/StatsThroughSkills/main.lua [hunk 1]
```diff
@@ -0,0 +1,393 @@
...
+        int_bonus = int_bonus + 1
+        if (computer_skill >= 6) then
+            int_bonus = int_bonus + 1
+            if (computer_skill >= 9) then
+                int_bonus = int_bonus + 1
+            end
+        end
+    end
+
+    cooking_skill = player:get_skill_level("cooking")
```
Comment (BevapDin) [inline data/mods/StatsThroughSkills/main.lua]: Am I the only one who things that all those blocks look identical and should go into a separate function?

```Lua
int_bonus = calc_skill_bonus(int_bonus, "cooking")
```
Reads far nicer and is easier to understand.

**Comment by DeNarr** [comment]
Ok, added in the functions, as well as the code the @kevingranade mentioned.

**Comment by Snaaty** [comment]
Does that mean that the range you can improve your stats on is from the base value of 8 up to 11?

**Comment by DeNarr** [comment]
@Snaaty No. Each skill can raise a stat by up to 3 points. With 7 skills for each stat, that would be a theoretical max of 29 if you managed to get all the skills to level 9. That would be quite difficult to do though, especially with skill rust.

**Comment by Sirbab** [comment]
If a skill degrades past the point that it gave you a stat boost, do you lose it? And more importantly, if not, is this abusable by raising and lowering a skill ad infinitum?

**Comment by DeNarr** [comment]
@Sirbab Yes, you will lost the stat point. All stat changes occur at midnight (since that's when the lua callback function is applied), and it will tell you which stats are lowered and raised.

**Comment by Malkeus** [comment]
How does MOD.on_day_passed() actually know when a day passes? I can't find it defined anywhere, it's driving me crazy.

**Comment by BevapDin** [comment]
> How does MOD.on_day_passed() actually know when a day passes?

A Lua function is called at midnight from game.cpp (`game::do_turn`). Each mod can attach itself to that callback, see `mod_callback` in lua/autoexec.lua.

**Comment by Malkeus** [comment]
Thank you.

**Comment by SatakAzat** [comment]
Hi, 
I know nothing in Lua but I'm having this message when midnight comes :
"DEBUG Lua returned unknown error 1 for mod_callback("on_day_passed")
lua/autoexec.lua:19: Unknown player attribute: get_skill_level
Press spacebar..."
It doesn't upped any stats with 10 meca, 5 computers, etc. Apparently the problem comes from having the skill's level but I don't know what to do, please help me !

**Comment by SatakAzat** [comment]
Hi,
I know nothing in Lua but I'm having this message when midnight comes :
"DEBUG Lua returned unknown error 1 for mod_callback("on_day_passed")
lua/autoexec.lua:19: Unknown player attribute: get_skill_level
Press spacebar..."
It doesn't upped any stats with 10 meca, 5 computers, etc. Apparently the problem comes from having the skill's level but I don't know what to do, please help me !

**Comment by Malkeus** [comment]
The current version of statsthruskills doesn't use the callback "on_day_passed" any longer, so I'm guessing you are using an outdated version of the mod. Possibly one of the alternate versions from the forum? I don't believe they've been updated as of yet.

**Comment by Malkeus** [comment]
Using the current version of statsthruskills with a slight change to the calc_bonus function will give results similar to the old overpowered one.
change "1/2.46" to "1/2.345" or (even lower if you want it to be more op) in calc_bonus's return line.

**Comment by SatakAzat** [comment]
You were right, I used an outdated version of the game, thanks for your help !

**Comment by DeNarr** [comment]
Hey, so I dropped out of the game for a bit, but decided to come back in and I'm rather surprised at the changes done to this mod. It was meant as an alternative way to raise your stats over time, and now it's essentially just getting a few free points in stats. 

Not putting any starting points into stats was a core aspect of the mod.
Meanwhile, it looks like if you got every single skill to 10 (which would be a crazy investment), you would only have +5 to your stats.

**Pull request merged**